### DISCUSSION TITLE
How to implement user authorization?

### DISCUSSION THREAD

**Discussion by caracal7** [uNetworking/uWebSockets.js#112] [Q&A] [answered]
How to implement reliable authorization? For example with passportJS or something like it.
And which is a proper way to manage user sessions with uWebSockets?

**Comment by Chathula** [comment]
+1

**Comment by mrlika** [comment]
It's a basic web server. It doesn't support middlewares like Passport.js. But you can implement almost everything on top of it. Read the documentation to understand how things work and go ahead!

**Comment by caracal7** [comment]
@mrlika 
I understand that it is a basic web server). And I implemented very simple authentication but I'm looking for the best practices and close to real world examples.
For now I simply send message from client with login/password after WebSocket opened and set `ws.logged = 'mylogin'` on server.
But something tell me that it is not a very safe and secure way)

**Comment by mrlika** [comment]
I don't think that the best practiced are related to this project. They are related to HTTP protocol in general.

**Comment by mrlika** [comment]
I think that the cookies are the way to go:
https://devcenter.heroku.com/articles/websocket-security#authentication-authorization

**Comment by girng** [comment]
> https://m.youtube.com/watch?v=OP30okjpCko
> 
> Best practise right there :)


😂😂🤣

**Comment by lagden** [comment]
@caracal7 and @Chathula 

Use cookie (`HttpOnly`)

```js
const {App} = require('uWebSockets.js')
const cookie = require('cookie')

const app = App()
app
  .ws('/*', {
    upgrade(res, req, context) {
      const _cookie = cookie.parse(req.getHeader('cookie'))
      // validate the cookie somehow
      // and set _logged true or false
      let _logged = true

      res.upgrade(
        {_logged},
        req.getHeader('sec-websocket-key'),
        req.getHeader('sec-websocket-protocol'),
        req.getHeader('sec-websocket-extensions'),
        context
      )
    },
    open(ws) {
      // disconnect if not logged
      if (!ws._logged) {
        ws.end()
        console.log('unauthorized', 'https://m.youtube.com/watch?v=OP30okjpCko')
        return
      }
    }
    // ...more code
  })
  .any('/*', (res, req) => {
    // set cookie
    res.writeHeader('Set-Cookie', '_token=jwt; SameSite=Strict; HttpOnly')
    res.end('cookie sample')
  })
  .listen('::', 3132, token => {
    console.info(token)
  })
```

**Comment by hst-m** [comment, marked as answer]
@lagden the point of the upgrade handler is to deny the request before the socket opens so you would want to update that

Here is JsonWebToken example
```javascript
app.ws('/', {
   upgrade:(res, req, context) => {
      try { res.user = decodeJwtCookie(res, req, 'cookieName'); }
      catch { return res.writeStatus('401').end(); }
      res.upgrade({ uid: res.user._id }, req.getHeader('sec-websocket-key'), req.getHeader('sec-websocket-protocol'), req.getHeader('sec-websocket-extensions'), context);
   },
   open: ws => console.log('open-ws', ws.uid)
});


const getCookie = (res, req, name) => { res.cookies ??= req.getHeader('cookie'); return res.cookies && res.cookies.match(getCookie[name] ??= new RegExp(`(^|;)\\s*${name}\\s*=\\s*([^;]+)`))?.[2]; };
const decodeJwtCookie = (res, req, name) => require('jsonwebtoken').verify(getCookie(res, req, name), env.jwtSecret);
```

`decodeJwtCookie` throws error if cookie was not signed with correct secret, and code prevents websocket from opening with `res.writeStatus('401').end()` unauthorized

**Reply by dragonlobster** [reply]
@hst-m is there an example of this using the async version of jwt.verify? i can't get it to work because the async function to get the user_id can't be assigned to res.upgrade context:
```
Error: Using uWS.HttpRequest past its request handler return is forbidden
```

**Reply by dragonlobster** [reply]
> @hst-m is there an example of this using the async version of jwt.verify? i can't get it to work because the async function to get the user_id can't be assigned to res.upgrade context:
> 
> ```
> Error: Using uWS.HttpRequest past its request handler return is forbidden
> ```

nevermind, i found the way, request object is not accessible after an async function so just need to store all the headers before the async function... this is an example with token in query parameter (probably NOT safe but just for demonstration):

```javascript
uWS.App()
.ws('/*', {

    upgrade: async (res, req, context) => {
                res.onAborted(() => {
                    res.aborted = true
                })
                try {
                    const token = req.getQuery("auth")

                    const key = req.getHeader("sec-websocket-key")
                    const protocol = req.getHeader("sec-websocket-protocol")
                    const extensions = req.getHeader("sec-websocket-extensions")
                    const sub = await verify(token)

                    if (res.aborted) return
                    if (!sub) return res.writeStatus("401").end()  

                    return res.upgrade({uid: sub}, key, protocol, extensions, context)
                }
        catch (err) {
        ...
```

**Reply by e3dio** [reply]
`onAborted` means the http connection is dead so `res.writeStatus` will throw error, instead set `res.aborted = true` and check `res.aborted` after async before calling `res.upgrade`

**Reply by dragonlobster** [reply]
thanks, i have made an edit

**Reply by e3dio** [reply]
Still not correct. Aborted means no more response. You are still trying to write status after aborted. This will throw error that you catch, but might as well not throw the error, return without writing response

**Reply by dragonlobster** [reply]
you are right, i ran into the error. second edit, i tried i think it's ok now

**Reply by e3dio** [reply]
`res.end()` is invalid, there is no more response after Aborted. Return without handling response `if (res.aborted) return;`

**Reply by dragonlobster** [reply]
third try's a charm

**Reply by Wyzix33** [reply]
I've used https://github.com/panva/paseto it was simple to implement, and it seems to be better and faster than jwt

**Reply by lagden** [reply]
I did the boilerplate project. It's very small, simple and has an auth example.
https://github.com/lagden/boilerplate-ws

I hope it helps someone. 😉
